### ISSUE TITLE
Pydantic Json[T] constructor input is modeled as T instead of raw JSON input

### ISSUE THREAD

**Issue by ChenMoFeiJin** [astral-sh/ty#3945]
### Summary

For a Pydantic field annotated as `Json[T]`, `ty` models the model constructor parameter as the field output type `T` rather than the raw JSON input accepted by Pydantic.

At runtime, `Json[T]` expects raw JSON input first:

```text
str | bytes | bytearray
```

The parsed JSON value is then validated as `T`.

Related to Pydantic `BaseModel` constructor signatures tracked in astral-sh/ty#2403.

### Reproducer

```python
from pydantic import BaseModel, ConfigDict, Json


class M(BaseModel):
    x: Json[list[int]]


class MS(BaseModel):
    model_config = ConfigDict(strict=True)
    x: Json[list[int]]


M(x="[1, 2]")      # runtime OK, ty accepts accidentally via Iterable[LaxInt]
M(x=b"[1, 2]")     # runtime OK, ty accepts accidentally via Iterable[LaxInt]
M(x=[1, 2])        # runtime error, ty accepts

MS(x="[1, 2]")     # runtime OK, ty rejects
MS(x=b"[1, 2]")    # runtime OK, ty rejects
MS(x=[1, 2])       # runtime error, ty accepts
```

### Actual behavior

With `ty 0.0.57`, the inferred constructor signatures are approximately:

```text
M.__init__:
(self: M, *, x: Iterable[LaxInt], **extra: Any) -> None

MS.__init__:
(self: MS, *, x: list[int], **extra: Any) -> None
```

This means lax mode accepts valid raw JSON inputs such as strings and bytes only accidentally because they are iterable. The inferred parameter type is still wrong, and it also accepts Python lists that Pydantic rejects for `Json[T]`.

Strict mode rejects valid raw JSON inputs:

```text
error[invalid-argument-type]: Argument is incorrect: Expected `list[int]`, found `Literal["[1, 2]"]`
error[invalid-argument-type]: Argument is incorrect: Expected `list[int]`, found `bytes`
```

### Expected behavior

The constructor parameter for `Json[T]` should be modeled as raw JSON input:

```text
x: str | bytes | bytearray
```

The field output type should still be `T`.

### Version

```text
ty 0.0.57
```

**Comment by sharkdp** [comment]
Pydantic [basically defines `Json[T]`](https://github.com/pydantic/pydantic/blob/cf67d4b3193c3fe43ede18612ed62785eee11382/pydantic/types.py#L1437-L1438) as `Json[T] = Annotated[T, ...]`, which means that it's essentially an alias for `T`. This is correct for the type of the field, but obviously wrong for the input type, which should probably be `str` in strict mode and `LaxStr = str | bytes | ...` in lax mode.

I'm not sure what to do about this. It would be great if this could be fixed on the Pydantic side. For example, I guess they could define `Json[T]` in their `if TYPE_CHECKING` block as a descriptor. Something like:

```py
if TYPE_CHECKING:
    class Json[T]:
        def __get__(self, instance, owner=None) -> T: ...
        def __set__(self, instance, value: str): ...    # or `value: str | bytes | ...`
```

This way, type checkers would understand that the input type is `str`, but the output type is `T` (ty demo with dataclasses:https://play.ty.dev/5252d43f-b382-4f38-ba66-9e9cf3f9c8b4):
```py
@dataclass
class Model:
    x: Json[list[int]]

m = Model(x="[1, 2]")  # good, accepts strings
reveal_type(m.x)  # good, returns list[int]

Model([1, 2])  # good, does not accept list[int]
```

FYI @Viicos
